cloudprovider.Zones should support external cloud providers

andrewsykim opened the pull request:
**What this PR does / why we need it**:
Provides methods in cloudprovider.Zones that allows external cloud providers to set the correct zone labels to nodes. Part of https://github.com/kubernetes/kubernetes/issues/48690

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: 
fixes https://github.com/kubernetes/kubernetes/issues/49308

**Special notes for your reviewer**:
Should help with getting ccm/external cloud providers to beta. 

**Release note**:
```release-note
cloudprovider.Zones should support external cloud providers
```

cc @luxas @wlan0 @thockin

Commit:
cloudprovider.Zones should support external cloud providers

andrewsykim commented on pkg/cloudprovider/cloud.go:
```diff
@@ -185,4 +185,14 @@ type Zone struct {
+	// GetZoneByNodeName returns the Zone containing the current zone and locality region of the node specified by node name
+	// This method is particularly used in the context of external cloud providers where node initialization must be down
+	// outside the kubelets.
+	GetZoneByNodeName(nodeName types.NodeName) (Zone, error)
```
Not a huge fan of having both `GetZoneByProviderID` and `GetZoneByNodeName`, but this is following a pattern in other cloud provider interfaces. Aside from having to maintain/support the methods, it does give external cloud providers flexibility in terms of how they would like to fetch node data (by id or name).

wlan0 replied:
@andrewsykim This is indeed not an ideal pattern, but like you mentioned it gives us the flexibility of choosing how we obtain zones from the cloud provider., either by name or id. We need both, simply because there are some nodes whose providerID may not be filled in - this scenario can occur on upgrades from older versions.

jhorwit2 commented on pkg/controller/cloud/node_controller.go:
```diff
@@ -430,3 +430,18 @@ func getInstanceTypeByProviderIDOrName(instances cloudprovider.Instances, node *
 	return instanceType, err
 }
+
+// getZoneByProviderIDorName will attempt to get the zone of node usint its providerID
```
s/usint/using

jhorwit2 commented on pkg/cloudprovider/providers/cloudstack/cloudstack.go:
```diff
@@ -130,3 +133,17 @@ func (cs *CSCloud) GetZone() (cloudprovider.Zone, error) {
 }
+
+// GetZoneByProviderID implements Zones.GetZoneByProviderID
+// This is particularly useful in external cloud providers where we can't rely on
```
I'd change this to say `where the kubelet does not initialize node data`.`rely` to me makes it seem like it's possible for the kubelet to do it for external providers, but it's not.

wlan0 commented on pkg/cloudprovider/cloud.go:
```diff
@@ -185,4 +185,14 @@ type Zone struct {
 type Zones interface {
 	// GetZone returns the Zone containing the current failure zone and locality region that the program is running in
 	GetZone() (Zone, error)
+
```
I would add a comment here saying that GetZone() originally was always run from the kubelet, and this method used the local metadata service on the node running in the cloudprovider to obtain this information. Therefore, there was no arguments needed to be passed in. 

Now that we are using an external cloud controller manager, that does not run on every node, we need to identify the node whose zone information we are obtaining. Hence there is a need for identifying the nodes in these specific ways.

wlan0 commented on pkg/controller/cloud/node_controller.go:
```diff
@@ -321,7 +321,7 @@ func (cnc *CloudNodeController) AddCloudNode(obj interface{}) {
 
 		if zones, ok := cnc.cloud.Zones(); ok {
-			zone, err := zones.GetZone()
+			zone, err := getZoneByProviderIDOrName(zones, curNode)
```
Thanks for fixing this. This originally always returned the zone in which the cloud controller manager was running.

FengyunPan replied:
+1
Do we update (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L310) too?

wlan0 commented:
Thanks for the PR @andrewsykim - This LGTM. There's just one more comment to add context for code readability.

wlan0 commented:
@andrewsykim Would you also be implementing the GetZoneByXYZ methods for any of the cloudproviders? 

That would help the other cloudproviders moving to this pattern by standing as an example.

andrewsykim commented:
@wlan0 thanks for reviewing. I've addressed your comment. 

I don't mind implementing the GetZoneBy methods for other cloud providers, just a matter of time for me. I'll open up an issue in case anyone else is willing to take up that work.

wlan0 commented:
/lgtm

Thank you @andrewsykim. Please CC me in the issue that you create.

andrewsykim commented:
@wlan0 issue opened https://github.com/kubernetes/kubernetes/issues/50926

wlan0 commented:
cc @thockin

FengyunPan commented:
/assign @mikedanese

andrewsykim commented:
/test pull-kubernetes-unit

andrewsykim commented:
/assign @thockin

andrewsykim commented:
/test pull-kubernetes-unit

thockin commented:
Who is going to implement this for the other providers?  Does this break CCM on all clouds?

andrewsykim commented:
@thockin we have a follow up issue https://github.com/kubernetes/kubernetes/issues/50926 to implement these methods for all the other clouds. The currenty implementation is already broken in that it returns the wrong zone information. This will "break" CCM only because the methods are not implemented.

thockin commented:
/lgtm
/approve

andrewsykim commented:
/test pull-kubernetes-unit

jhorwit2 commented:
fyi @andrewsykim the fix for that flake is here https://github.com/kubernetes/kubernetes/pull/51258

andrewsykim commented:
@jhorwit2 thought it was https://github.com/kubernetes/kubernetes/pull/51204

FengyunPan commented:
/lgtm

dixudx commented:
/retest

FengyunPan commented:
/retest

luxas commented:
@andrewsykim Thanks! Did you file an issue for populating these methods for all in-tree providers?
That's the next step of this efffort...

andrewsykim commented:
@luxas https://github.com/kubernetes/kubernetes/issues/50926

cheftako commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -1201,6 +1201,20 @@ func (c *Cloud) GetZone() (cloudprovider.Zone, error) {
+// This is particularly useful in external cloud providers where the kubelet
+// does not initialize node data.
+func (c *Cloud) GetZoneByNodeName(nodeName types.NodeName) (cloudprovider.Zone, error) {
+	return cloudprovider.Zone{}, errors.New("GetZoneByNodeName not imeplemented")
```
implemented

cheftako commented on pkg/cloudprovider/providers/azure/azure_zones.go:
```diff
@@ -55,6 +57,20 @@ func (az *Cloud) GetZone() (cloudprovider.Zone, error) {
+// This is particularly useful in external cloud providers where the kubelet
+// does not initialize node data.
+func (az *Cloud) GetZoneByNodeName(nodeName types.NodeName) (cloudprovider.Zone, error) {
+	return cloudprovider.Zone{}, errors.New("GetZoneByNodeName not imeplemented")
```
implemented

cheftako commented on pkg/cloudprovider/providers/cloudstack/cloudstack.go:
```diff
@@ -130,3 +133,17 @@ func (cs *CSCloud) GetZone() (cloudprovider.Zone, error) {
+// This is particularly useful in external cloud providers where the kubelet
+// does not initialize node data.
+func (cs *CSCloud) GetZoneByNodeName(nodeName types.NodeName) (cloudprovider.Zone, error) {
+	return cloudprovider.Zone{}, errors.New("GetZoneByNodeName not imeplemented")
```
implemented
